controllers/cloudstackfailuredomain_controller_test.go

```diff
@@ -49,14 +48,83 @@ var _ = Describe("CloudStackFailureDomainReconciler", func() {
-				key := client.ObjectKeyFromObject(dummies.CSFailureDomain1)
-				if err := k8sClient.Get(ctx, key, tempfd); err == nil {
-					return tempfd.Status.Ready
-				}
-				return false
-			}, timeout).WithPolling(pollInterval).Should(BeTrue())
 		})
+
+		It("Should set failure domain Status.Ready to true.", func() {
+			assertFailureDomainCreated()
+		})
+		It("Should delete failure domain if no VM under this failure domain.", func() {
+			assertFailureDomainCreated()
+			Ω(k8sClient.Delete(ctx, dummies.CSFailureDomain1))
+
+			assertFailureDomainNotExisted()
+		})
+		DescribeTable("Should function in different replicas conditions",
+			func(shouldDeleteVM bool, specReplicas, statusReplicas, statusReadyReplicas *int32, statusReady *bool) {
+				assertFailureDomainCreated()
+				setCSMachineOwnerCRD(dummies.CSMachineOwner, specReplicas, statusReplicas, statusReadyReplicas, statusReady)
+				setCAPIMachineAndCSMachineCRDs(dummies.CSMachine1, dummies.CAPIMachine)
+				setMachineOwnerReference(dummies.CSMachine1, dummies.CSMachineOwnerReference)
+				labelMachineFailuredomain(dummies.CSMachine1, dummies.CSFailureDomain1)
+
+				Ω(k8sClient.Delete(ctx, dummies.CSFailureDomain1))
+
+				CAPIMachine := &clusterv1.Machine{}
+				Eventually(func() bool {
+					key := client.ObjectKey{Namespace: dummies.ClusterNameSpace, Name: dummies.CAPIMachine.Name}
+					if shouldDeleteVM {
+						if err := k8sClient.Get(ctx, key, CAPIMachine); err != nil {
+							return errors.IsNotFound(err)
+						}
+					} else {
+						if err := k8sClient.Get(ctx, key, CAPIMachine); err == nil {
+							return CAPIMachine.DeletionTimestamp.IsZero()
+						}
+					}
+					return false
+				}, timeout).WithPolling(pollInterval).Should(BeTrue())
+			},
+			// should delete - simulate owner is kubeadmcontrolplane
+			Entry("Should delete machine if spec.replicas > 1", true, int32Pointer(2), int32Pointer(2), int32Pointer(2), boolPointer(true)),
+			// should delete - simulate owner is etcdadmcluster
+			Entry("Should delete machine if status.readyReplica does not exist", true, int32Pointer(2), int32Pointer(2), nil, boolPointer(true)),
+			// should delete - simulate owner is machineset
+			Entry("Should delete machine if status.ready does not exist", true, int32Pointer(2), int32Pointer(2), int32Pointer(2), nil),
+			// should not delete if condition not met
+			Entry("Should return error if status.replicas < spec.replicas", false, int32Pointer(2), int32Pointer(1), int32Pointer(1), boolPointer(true)),
+			Entry("Should return error if spec.replicas < 2", false, int32Pointer(1), int32Pointer(1), int32Pointer(1), boolPointer(true)),
+			Entry("Should return error if status.ready is false", false, int32Pointer(2), int32Pointer(2), int32Pointer(2), boolPointer(false)),
+			Entry("Should return error if status.readyReplicas <> status.replicas", false, int32Pointer(2), int32Pointer(2), int32Pointer(1), boolPointer(true)),
+		)
 	})
 })
+
+func assertFailureDomainCreated() {
```

Comment:
I'm fine with this as is, but my preference would be to wrap the getter and checker in an Eventually outside of a function like this. That way the eventually is more obvious when reading through a spec.

This kind of hides the eventually in a method and makes it slightly more opaque that the code may just hang at this spot until an eventual failure occurs.